server/api_archive.go

```diff
@@ -0,0 +1,197 @@
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	znw := zip.NewWriter(buffer)
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only while happening in a single write. We must archive the files in reverse order from
+	// their dependency order.
+
+	for _, name := range PublicFiles {
+		err := gcas.addFile(name, znw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Add the pseudo file server.pubkey last, as the other files are signed by it.
+	if err := gcas.addPubKeyFile("server.pubkey", znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Add a README
+	if err := gcas.addReadmeFile(znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := znw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
+}
+
+// Add a file to a zip archive, copying all the data in the file.
+func (gcas *GCAServer) addFile(name string, zipw *zip.Writer) error {
+	path := filepath.Join(gcas.baseDir, name)
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := zip.FileInfoHeader(fileInfo)
+	if err != nil {
+		return err
+	}
+
+	header.Method = zip.Deflate
+
+	writer, err := zipw.CreateHeader(header)
+	if err != nil {
+		return err
+	}
+
+	// File contains only public data so copy it all.
+	if _, err := io.Copy(writer, file); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Add a pseudo-file "server.pubkey" to a zip archive,
+// copying only the public key part of the data file "server.keys".
+func (gcas *GCAServer) addPubKeyFile(name string, znw *zip.Writer) error {
+
+	path := filepath.Join(gcas.baseDir, "server.keys")
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	// Read only the public key from the file, using the
+	// server API.
+	pub, _, err := gcas.loadGCAServerKeys()
+	if err != nil {
+		return err
+	}
+
+	reader := bytes.NewReader(pub[:])
+
+	// Read the first 32 bytes from the server.keys file
+	// and verify that we have the same public key data.
+
+	buf := make([]byte, 32)
+
+	if _, err := file.Read(buf); err != nil {
+		return err
+	}
+
```

Comment:
extra newline - in general go is verbose enough that you don't need too many newlines. There's no strict rule but I usually aim for somewhere between 5 and 12 lines of code between each newline. Really, I try to use newlines to separate major code ideas. Another sign that you don't need newlines between consecutive lines of code is if you have a comment above them that applies to all following lines.

It's not a hard science, in general though your code could trim out probably around half the newlines you put in. This one stretch here just particularly stood out because all of the ideas are closely associated with each other and there's very low complexity to the chain of events.

Reply:
This is a style thing. I tend to use more newlines in any language, as I find the code is more readable that way. However I'm fine being more terse about it.